onnxruntime/test/providers/provider_test_utils.h

```diff
@@ -462,15 +506,30 @@ class OpTester {
   void FillFeedsAndOutputNames(std::unordered_map<std::string, OrtValue>& feeds,
                                std::vector<std::string>& output_names);
 
-  std::unique_ptr<onnxruntime::Model> BuildGraph();
+  template <class SessionType>
+  std::vector<MLValue> ExecuteModel(Model& model,
+                                    SessionType& session_object,
+                                    ExpectResult expect_result,
+                                    const std::string& expected_failure_string,
+                                    const RunOptions* run_options,
+                                    std::unordered_map<std::string, OrtValue> feeds,
+                                    std::vector<std::string> output_names,
+                                    const std::string& provider_type,
+                                    const CustomOutputVerifierFn& custom_output_verifier);
 
   const char* op_;
+  std::vector<Data> input_data_;
+  std::vector<Data> output_data_;
+  std::vector<OrtValue> fetches_;
+
+  // for gradient unit tests only 
+  std::shared_ptr<onnxruntime::Model> cached_model_;
 
 #ifndef NDEBUG
   bool run_called_{};
 #endif
```

Comment:
Can these be private as non-private data members should generally be avoided?

Reply:
Agreed. There's code change on in 3/2019 that move the fields from private to protected. I think it's for the convenience/shortcut of unit tests.

Reply:
Unfortunately new developers might see these sort of convenience shortcuts, think that's the best way to do things, and they get replicated in production code. Hopefully they can be made private in the near future as the cost of adding some accessors is pretty low and makes it much easier to debug if something weird happens as you can just put a breakpoint on the accessor to find all places that use the variable, vs. having to search all the code to find all the places in derived classes that directly access them because they're protected. 

---
In reply to: [377998095](https://github.com/microsoft/onnxruntime/pull/2999#discussion_r377998095) [](ancestors = 377998095)